### PR TITLE
[ENH] Refactor sliding and expanding window splitters to allow timdelta forecasting horizon

### PR DESCRIPTION
**Pull request by khrapovs**
#### Reference Issues/PRs

Next step in #1737

#### What does this implement/fix? Explain your changes.

In general, I have refactored splitters (mainly `SlidingWindowSplitter` and `ExpandingWindowSplitter`) so that they accept `pd.Timedelta` as forecasting horizons.

More detailed list of changes:
- Conditions in `_check_window_lengths` are simplified taken into account that we now support only two options: either everything (windows, horizons, etc.) is integer, or everything is timdelta-like.
- `get_cutoffs` now returns only integer arrays. This is a unification step to make it clear that we return `k` instead of either `k` or `t(k)`. All implementations, docstrings and test function `_check_cutoffs_against_train_windows` are updated accordingly.
- New method `_split_windows_generic` in `BaseWindowSplitter`. It encapsulates code repetition from two implementations `_split_windows` in `SlidingWindowSplitter` and `ExpandingWindowSplitter`. The difference is regulated by boolean argument `expanding`.
- Most importantly usage of `_coerce_duration_to_int` with hardcoded `freq="D"` is gone. Now we do not care if timedelta is daily, weekly, or any other frequency.
- All test functions for `SlidingWindowSplitter` and `ExpandingWindowSplitter` have additional test cases `TEST_FHS_TIMEDELTA`.

#### Does your contribution introduce a new dependency? If yes, which one?

No

##### For all contributions
- [X] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [X] I've added unit tests and made sure they pass locally.
- [X] The PR title starts with either [ENH], [DOC] or [BUG] indicating whether the PR topic is related to enhancement, documentation or bug

### CODE REVIEW

**Comment by khrapovs** [comment]
@fkiraly @aiwalter Sorry to bother you, but would you be so kind to take a look at this PR? Thanks!

**Comment by fkiraly** [comment]
sorry to let you wait, forgot this over GSoC and 0.11.3 and all. Sure!

**Comment by fkiraly** [comment]
looks ok, let me just quickly test whether it would break in combination with #2599

**Comment by khrapovs** [comment]
Thanks for suggestions! As you requested, I have updated dosctrings for `_split_windows_generic` and for `get_cutoffs`. I made it even more explicit that `get_cutoffs` always return `.iloc[]` integer indices.

File: sktime/forecasting/model_selection/_split.py [hunk 1]
```diff
@@ -472,8 +471,8 @@ def get_cutoffs(self, y: Optional[ACCEPTED_Y_TYPES] = None) -> np.ndarray:
 
         Returns
         -------
-        cutoffs : np.array
-            The array of cutoff points.
+        cutoffs : np.ndarray
+            The array of integer cutoff points.
```
Comment (fkiraly) [inline sktime/forecasting/model_selection/_split.py:475]: I would suggest to write:
1D np.ndarray of `int`
`iloc` location indices, in reference to `y`, of cutoff indices
(is that correct?)

File: sktime/forecasting/model_selection/_split.py [hunk 2]
```diff
@@ -707,41 +700,81 @@ def _split_for_initial_window(self, y: pd.Index) -> SPLIT_ARRAY_TYPE:
...
+            Index of time series to split
+        fh : ForecastingHorizon
+            Single step ahead or array of steps ahead to forecast.
+        expanding : bool
+            Expanding (True) or sliding window (False) splitter
+
+        Yields
+        ------
+        training_window : np.ndarray
+            Training window indices
```
Comment (fkiraly) [inline sktime/forecasting/model_selection/_split.py:749]: same here. Are these `iloc` or `loc`references? And, reference, to what?

**Review by fkiraly — CHANGES_REQUESTED** [review]
Excellent!

Could you just be a little bit more specific?
Made suggestions, but I´m not sure whether they are correct.

File: sktime/forecasting/model_selection/_split.py (hunk 1 above)
Comment (khrapovs) [reply sktime/forecasting/model_selection/_split.py]: yes, totally agree. changed in all 3 palces

**Comment by khrapovs** [comment]
> Could you just be a little bit more specific? Made suggestions, but I´m not sure whether they are correct.

All suggestions are fine with me. Tests pass. Anything else I can do here?

**Comment by fkiraly** [comment]
you said you changed it in 3 places, but only 1 place is changed. Probably you just forgot to push.

**Comment by khrapovs** [comment]
these two lines:
```
cutoffs : 1D np.ndarray of int
            iloc location indices, in reference to y, of cutoff indices
```
can be found in line 474, 631, 853, and 1071. You are right, I forgot to push one. Thanks.

**Comment by fkiraly** [comment]
I think it´s still missing in line 749, 751? Did you push everything?

**Comment by khrapovs** [comment]
> I think it´s still missing in line 749, 751? Did you push everything?

Lines 748-751:
```
train : 1D np.ndarray of int
    Training window iloc indices, in reference to y
test : 1D np.ndarray of int
    Test window iloc indices, in reference to y
```

**Review by fkiraly — APPROVED** [review]
Thanks! all good now.

**Pull request merged**